### PR TITLE
[money] Add types definitions for money

### PR DESCRIPTION
**Pull request by hozblok**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## money ([<kbd>unpkg</kbd>](https://unpkg.com/browse/money@latest/))
was missing the following properties:
1. noConflict
1. fx
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 496efeeb0ca6b96cf6b00b139dcaac6d44acf8d9
</p>

File: types/money/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    convert: {
+        <T_VAL extends fx.Value>(val: T_VAL, opts?: fx.Options): fx.Result<T_VAL>;
+    };
+}
+
+/**
+ * Library for realtime currency conversion and exchange rate calculation
+ */
+declare const fx: FX;
+export = fx;
```
Comment (peterblazejewicz) [inline types/money/index.d.ts:87]:
```suggestion

export as namespace fx;

export = fx;
```
this is UMD module:
https://github.com/openexchangerates/money.js/blob/master/money.js#L133
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-a-umd-library

File: types/money/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /**
+     * If fx(val) is called as a function, it returns a wrapped object that can be used OO-style
+     */
+    <T_VAL extends fx.Value>(val: T_VAL): fx.Wrapper<T_VAL>;
+    new <T_VAL extends fx.Value>(val: T_VAL): fx.Wrapper<T_VAL>;
+
+    /**
+     * Current version
+     */
+    version: string;
```
Comment (peterblazejewicz) [inline types/money/index.d.ts:55]:
```suggestion
    version: '0.2';
```
IMO type literal works best here (for versions)

File: types/money/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        from: fx.Currency;
+        to: fx.Currency;
+    };
+
+    /**
+     * Converts a value from one currency to another
+     */
+    convert: {
+        <T_VAL extends fx.Value>(val: T_VAL, opts?: fx.Options): fx.Result<T_VAL>;
+    };
```
Comment (peterblazejewicz) [inline types/money/index.d.ts:80]:
```suggestion
    convert: <T_VAL extends fx.Value>(val: T_VAL, opts?: fx.Options) => fx.Result<T_VAL>;
```
optionally, for readability

File: types/money/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        };
+
+        to: {
+            (currency: Currency): Result<T_VAL>;
+        };
+    }
+
+    interface Wrapper<T_VAL extends Value> extends WrapperPrototype<T_VAL> {
+        readonly _v: number;
+        readonly _fx?: Currency;
```
Comment (peterblazejewicz) [inline types/money/index.d.ts:36]: those are private, what's the purpose when consuming this types?

File: types/money/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+declare namespace fx {
+    type Currency = string;
+
+    type Rates = Record<string, number | string>;
+
+    type Options = {
+        from?: Currency | null;
+        to?: Currency | null;
+    } | null;
```
Comment (peterblazejewicz) [inline types/money/index.d.ts:14]:
```suggestion
    } ;
```
IMO just set options param as `options?` (optional), there is no trace of explicit null, just expression check for undefined/nullish options:
https://github.com/openexchangerates/money.js/blob/master/money.js#L59

**Review by gabritto — CHANGES_REQUESTED** [review]
Please address the comments left by Peter.

File: types/money/index.d.ts (hunk 4 above)
Comment (hozblok) [reply types/money/index.d.ts:36]: I viewed it as protected not as private. The case is the user want to inherit the NewableFunction `fx`:

```typescript
import fx = require('money');

class MyFX<T_VAL extends fx.Value> extends fx<T_VAL> {
    constructor(val: T_VAL) {
        super(val);
    }

    getValue() {
        // ... do something
        return this._v // number
    }
}
```

Maybe remove `readonly`... How do you think?

File: types/money/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        readonly _fx?: Currency;
+    }
+
+    interface WrapperConstructor<T_VAL extends Value> {
+        readonly prototype: WrapperPrototype<T_VAL>;
+        new (val: T_VAL): Wrapper<T_VAL>;
+    }
+}
+
+interface FX {
```
Comment (hozblok) [inline types/money/index.d.ts:45]: It's better to put FX inside the namespace.

File: types/money/index.d.ts (hunk 2 above)
Comment (hozblok) [reply types/money/index.d.ts:55]: What will happen if the version of the library is updated to `'0.2.1'`? And no code changes with respect to types...

Will our user have to search for the required version of the `@types/money`? or put update to `DefinitelyTyped/DefinitelyTyped`? Version `'0.2.1'` in the variable with `'0.2'` type looks confusing.
How do you think?

File: types/money/index.d.ts (hunk 4 above)
Comment (peterblazejewicz) [reply types/money/index.d.ts]: don't be mislead by the power of types. This is a module and I think it only exports a single symbol, which is an instance of those constructs. There is no way for user to extend something, am I seeing this right?

Comment (hozblok) [reply types/money/index.d.ts]: Oh, you are right. Sorry for misunderstanding.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@hozblok thanks for your first contribution!

**Comment by hozblok** [comment]
@peterblazejewicz Thanks a lot for the review! I am inspired by TS, monaco and vscode. So if there is a priority task to contribute to the community, I can do) my email: hozblok@gmail.com

**Comment by hozblok** [comment]
Ready to merge

**Pull request merged**